pkg/fetcher/fetcher_test.go

```diff
@@ -0,0 +1,103 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetcher
+
+import (
+	"bytes"
+	"io/ioutil"
+	"testing"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	errJsonStr401 = `{
+	"errors":
+		[{"code":"UNAUTHORIZED",
+		  "message":"authentication required",
+		  "detail":[{"Type":"repository","Class":"","Name":"library/jiowengew","Action":"pull"}]
+		}]
+	}
+	`
+	multipleErrJsonStr = `{
+	"errors":
+		[{"code":"UNAUTHORIZED",
+		  "message":"authentication required",
+		  "detail":[{"Type":"repository","Class":"","Name":"library/jiowengew","Action":"pull"}]
+		},
+		{"code":"NOTFOUND",
+		 "message": "image not found",
+		 "detail": "image not found"
+		}]
+	}
+	`
+	unexpectedStr                = `random`
+	unexpectedJsonStr            = `{"nope":"nope"}`
+	errJsonWithEmptyErrorsField  = `{"errors":[]}`
+	errJsonWithNoMessageField    = `{"errors":[{"code":"nope","detail":"nope"}]}`
+	errJsonWithEmptyMessageField = `{"errors":[{"code":"nope","message":""},{"message":""}]}`
+)
+
+func TestExtractErrResponseMessage(t *testing.T) {
+	// Test set up: create the io streams for testing purposes
+	// multiple streams needed: these streams only have read ends
+	singleErrTestStream := ioutil.NopCloser(bytes.NewReader([]byte(errJsonStr401)))
+	multipleErrTestStream := ioutil.NopCloser(bytes.NewReader([]byte(multipleErrJsonStr)))
+	unexpectedStrTestStream := ioutil.NopCloser(bytes.NewReader([]byte(unexpectedStr)))
+	malformedJSONTestStream := ioutil.NopCloser(bytes.NewReader([]byte(unexpectedJsonStr)))
+	emptyErrorsJSONTestStream := ioutil.NopCloser(bytes.NewReader([]byte(errJsonWithEmptyErrorsField)))
+	noMessageJSONTestStream := ioutil.NopCloser(bytes.NewReader([]byte(errJsonWithNoMessageField)))
+	emptyMessageJSONTestStream := ioutil.NopCloser(bytes.NewReader([]byte(errJsonWithEmptyMessageField)))
+
+	// Test 1: single error message extraction
+	msg, err := extractErrResponseMessage(singleErrTestStream)
+	assert.Nil(t, err, "test: (single error message) extraction should success for well-formatted error json")
+	assert.Equal(t, "authentication required", msg,
+		"test: (single error message) extracted message: %s; expected: authentication required", msg)
+
+	// Test 2: multiple error message extraction
+	msg, err = extractErrResponseMessage(multipleErrTestStream)
+	assert.Nil(t, err, "test: (multiple error messages) extraction should success for well-formatted error json")
+	assert.Equal(t, "authentication required, image not found", msg,
+		"test: (multiple error messages) extracted message: %s; expected: authentication required, image not found", msg)
+
+	// Test 3: random string in the stream that is not a json
+	msg, err = extractErrResponseMessage(unexpectedStrTestStream)
+	assert.Equal(t, "", msg, "test: (non-json string) no message should be extracted")
+	assert.NotNil(t, err, "test: (non-json string) extraction should fail")
```

Comment:
Instead of `assert.NotNil(t, err, "test: (non-json string) extraction should fail")`, we should have something like `assert.Equal(t, err, errJsonFormat, "test: (non-json string) extraction should fail")`.